### PR TITLE
Adding capability for reading configuration knobs from host OS

### PR DESCRIPTION
**Pull request by kuqin12**
# Preface

Please ensure you have read the [contribution docs](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md) prior
to submitting the pull request. In particular,
[pull request guidelines](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md#pull-request-best-practices).

## Description

This change added the capability to read configuration variables into specified variable list format to allow validation engineers to verify the applied configuration knobs takes effect.

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

This is tested on local workstations.

## Integration Instructions

Developer could specify configuration definition using `-x` to allow the script to only look up xml related variables. `-o` is also specified to allow user to change the output file name, otherwise "output.vl" is the default.

### CODE REVIEW

File: SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
+# @file
+#
+# Set a dmpstore formatted set of variables to NVRAM
```
Comment (kuqin12) [inline SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py:3]: Update comments for new file

File: SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    if arguments.configuration_file is not None and not os.path.isfile(arguments.configuration_file):
+        print("Invalid input file: %s" % arguments.configuration_file)
+        sys.exit(1)
+
+    return arguments
+
+
+#
+# Create an unpack statement formatted to address the variable length
```
Comment (kuqin12) [inline SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py:51]: Update function description

File: SetupDataPkg/Tools/SettingSupport/UefiVariablesSupportLib.py [hunk 3]
```diff
@@ -123,6 +134,41 @@ def GetUefiVar(self, name, guid):
             return (err, None, WinError(err))
         return (err, efi_var[:length], None)
 
+    #
+    # Function to get variable
```
Comment (kuqin12) [inline SetupDataPkg/Tools/SettingSupport/UefiVariablesSupportLib.py:138]: Update function description

File: SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py (hunk 1 above)
Comment (kuqin12) [reply SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py]: Updated.

File: SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py (hunk 2 above)
Comment (kuqin12) [reply SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py]: Updated.

**Review by apop5 — APPROVED** [review]
This looks okay to me.

File: SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                logging.error(f"Error returned from GetUefiAllVarNames: {rc}")
+
+            offset = 0
+            while offset < len(efi_var_names):
+                (next_offset,) = struct.unpack_from("<I", efi_var_names[offset:])
+                if next_offset == 0:
+                    # This is the end... But we still need to go through the last loop
+                    next_offset = len(efi_var_names) - offset
+                namespace = uuid.UUID(bytes_le=efi_var_names[offset + 0x04: offset + 0x14])
+                name = efi_var_names[offset + 0x14: offset + next_offset].decode('utf16')
```
[the remark this answers is not in the document]
Comment (kuqin12) [reply SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py:90]: Yep, will do

[the remark this answers is not in the document]
Comment (kuqin12) [reply SetupDataPkg/Tools/ReadUefiVarsToConfVarList.py]: Updated.

**Pull request merged**